In [1]:
import pandas as pd
import os
import csv
pd.set_option('display.max_columns', 60)
#pd.set_option('display.max', 40)
import warnings

In [2]:
rd = pd.read_csv("C:/Users/GiridharRavindran/Desktop/Data/gpi/ndw_v_product.txt", sep='|',index_col=False,converters={'NDC':str,'CMF_PACK_NBR':str,'GPI10_CD':str,'UPC':str})

C:\Users\GiridharRavindran\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (54) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
rd.head(10)

PRODUCT_ID          NDC   CMF_PPK_NBR  CMF_PROD_NBR CMF_PACK_NBR  \
0     1368418  00067434790  1.621283e+09     1621283.0          010   
1     1478043  86790001960           NaN           NaN                
2     1368417  00067434760  1.621283e+09     1621283.0          010   
3     1382655  00067434772  1.621283e+09     1621283.0          010   
4     1478050  86790001930  1.999980e+09     1999980.0          001   
5     1368416  00067434730  1.621283e+09     1621283.0          010   
6     1011934  00083012530  1.621283e+09     1621283.0          001   
7     1008379  00067012575  1.621283e+09     1621283.0          001   
8     1008378  00067012547  1.621283e+09     1621283.0          001   
9     1011933  00083012417  1.621287e+09     1621287.0          001   

                               MKTED_PROD_FORMLTN_NM  \
0        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
1        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
2        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
3        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
4        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
5        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
6        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG   
7        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG   
8        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG   
9  SLOW FE/FOLIC ACID ORAL TABLET SR 24 HR 160-0....   

                         MKTED_PROD_FORMLTN_SHORT_NM      MKTED_PROD_DETL_NM  \
0        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
1        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
2        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
3        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
4        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
5        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
6        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG             SLOW FE OTC   
7        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG             SLOW FE OTC   
8        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG             SLOW FE OTC   
9  SLOW FE/FOLIC ACID ORAL TABLET SR 24 HR 160-0....  SLOW FE/FOLIC ACID OTC   

  MKTED_PROD_TYP_CD       MKTED_PROD_NM             PACK_LABEL_NM  \
0                 T             SLOW FE         SLOW FE TAB 142MG   
1                 T             SLOW FE          SLOW FE TAB 45MG   
2                 T             SLOW FE         SLOW FE TAB 142MG   
3                 T             SLOW FE         SLOW FE TAB 142MG   
4                 T             SLOW FE          SLOW FE TAB 45MG   
5                 T             SLOW FE         SLOW FE TAB 142MG   
6                 T             SLOW FE       SLOW FE TAB 50MG CR   
7                 T             SLOW FE      SLOW FE TAB 160MG CR   
8                 T             SLOW FE      SLOW FE TAB 160MG CR   
9                 T  SLOW FE/FOLIC ACID  FE/FOLIC ACI TAB 160-0.4   

                UNIFM_PROD_NM DOSAGE_FORM_CD           DOSAGE_FORM_NM  \
0             FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
1             FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
2             FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
3             FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
4             FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
5             FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
6       FERROUS SULFATE DRIED           TBCR  TABLET EXTENDED RELEASE   
7       FERROUS SULFATE DRIED           TBCR  TABLET EXTENDED RELEASE   
8       FERROUS SULFATE DRIED           TBCR  TABLET EXTENDED RELEASE   
9  FERROUS SULFATE-FOLIC ACID           TB24          TABLET SR 24 HR   

        STRNT_DESC  STRNT_VALUE_NBR STRNT_UOM_CD  PACK_QTY PACK_QTY_UOM_CD  \
0           142 MG            142.0           MG       1.0             BOX   
1  

Get datatype of all the columns

In [4]:
print(rd.dtypes)

PRODUCT_ID                       int64
NDC                             object
CMF_PPK_NBR                    float64
CMF_PROD_NBR                   float64
CMF_PACK_NBR                    object
MKTED_PROD_FORMLTN_NM           object
MKTED_PROD_FORMLTN_SHORT_NM     object
MKTED_PROD_DETL_NM              object
MKTED_PROD_TYP_CD               object
MKTED_PROD_NM                   object
PACK_LABEL_NM                   object
UNIFM_PROD_NM                   object
DOSAGE_FORM_CD                  object
DOSAGE_FORM_NM                  object
STRNT_DESC                      object
STRNT_VALUE_NBR                float64
STRNT_UOM_CD                    object
PACK_QTY                       float64
PACK_QTY_UOM_CD                 object
PACK_SIZE_NBR                  float64
PACK_SIZE_UOM_CD                object
TOTAL_PACK_QTY                 float64
ROUTE_ADM_CD                    object
ROUTE_ADM_NM                    object
GLBL_ROUTE_FORM_CD              object
UPC                      

In [5]:
# Fill nan with zeros 
rd.fillna(0, inplace=True)
rd.head(1)

PRODUCT_ID          NDC   CMF_PPK_NBR  CMF_PROD_NBR CMF_PACK_NBR  \
0     1368418  00067434790  1.621283e+09     1621283.0          010   

                         MKTED_PROD_FORMLTN_NM  \
0  SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   

                   MKTED_PROD_FORMLTN_SHORT_NM MKTED_PROD_DETL_NM  \
0  SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG        SLOW FE OTC   

  MKTED_PROD_TYP_CD MKTED_PROD_NM      PACK_LABEL_NM    UNIFM_PROD_NM  \
0                 T       SLOW FE  SLOW FE TAB 142MG  FERROUS SULFATE   

  DOSAGE_FORM_CD           DOSAGE_FORM_NM STRNT_DESC  STRNT_VALUE_NBR  \
0           TBCR  TABLET EXTENDED RELEASE     142 MG            142.0   

  STRNT_UOM_CD  PACK_QTY PACK_QTY_UOM_CD  PACK_SIZE_NBR PACK_SIZE_UOM_CD  \
0           MG       1.0             BOX           90.0               EA   

   TOTAL_PACK_QTY ROUTE_ADM_CD ROUTE_ADM_NM GLBL_ROUTE_FORM_CD          UPC  \
0            90.0           OR         ORAL                BAA  30067434790   

  HCPCS_CD HCPCS_DESC  DEA_CLAS_CD                DEA_CLAS_DESC  ...  \
0        0          0          0.0  DEA Class is not applicable  ...   

   OBSOLETE_DT  ACT_IND   USC_CD                USC_DESC  USC_LVL4_CD  \
0            0        A  48111.0  HEMATINICS, IRON ALONE      48110.0   

         USC_LVL4_DESC  USC_LVL3_CD USC_LVL3_DESC USC_LVL2_CD  USC_LVL2_DESC  \
0  HEMATINICS, FERROUS      48100.0    HEMATINICS     48000.0  BLOOD FACTORS   

              GPI                                           GPI_DESC  \
0  82300010000455  Ferrous Sulfate Tab CR 142 MG (45 MG Fe Equiva...   

     GPI10_CD       GPI10_DESC   GPI6_CD GPI6_DESC  GPI4_CD GPI4_DESC  \
0  8230001000  FERROUS SULFATE  823000.0   IRON***   8230.0    IRON**   

   GPI2_CD              GPI2_DESC  GPI_THPTC_CLAS_ID   GPI_THPTC_CLAS_DESC  \
0     82.0  HEMATOPOIETIC AGENTS*               15.0  HEMATOLOGICAL AGENTS   

   GENERIC_THPTC_CLAS_ID GENERIC_THPTC_CLAS_DESC     AHFS_CD  \
0                   59.0       ELECT/CALORIC/H2O  20040400.0   

           AHFS_DESC THPTC_CLAS_ID                      THPTC_CLAS_DESC  \
0  IRON PREPARATIONS          88.0  HEMATINICS & BLOOD CELL STIMULATORS   

   CMF_MKTED_PROD_NM CMF_USC_CD  
0            SLOW FE    48111.0  

[1 rows x 74 columns]

Change datatype of numbered columns for later computation

In [6]:
#rd['CMF_PACK_NBR'] = rd.CMF_PACK_NBR.astype(int)
rd['CMF_PPK_NBR'] = rd.CMF_PPK_NBR.astype(int)
rd['CMF_PROD_NBR'] = rd.CMF_PROD_NBR.astype(int)
rd['CMF_USC_CD'] = rd.CMF_USC_CD.astype(int)
rd['THPTC_CLAS_ID'] = rd.THPTC_CLAS_ID.astype(int)
rd['AHFS_CD'] = rd.AHFS_CD.astype(int)
rd['GENERIC_THPTC_CLAS_ID'] = rd.GENERIC_THPTC_CLAS_ID.astype(int)
rd['GPI_THPTC_CLAS_ID'] = rd.GPI_THPTC_CLAS_ID.astype(int)
#rd['GPI10_CD'] = rd.GPI10_CD.astype(int)
rd['GPI2_CD'] = rd.GPI2_CD.astype(int)
rd['GPI4_CD'] = rd.GPI4_CD.astype(int)
rd['GPI6_CD'] = rd.GPI6_CD.astype(int)
rd['USC_LVL2_CD'] = rd.USC_LVL2_CD.astype(int)
rd['USC_LVL3_CD'] = rd.USC_LVL3_CD.astype(int)
rd['USC_LVL4_CD'] = rd.USC_LVL4_CD.astype(int)
rd['USC_CD'] = rd.USC_CD.astype(int)
#rd['UPC'] = rd.UPC.astype(int)
rd['STRNT_VALUE_NBR'] = rd.STRNT_VALUE_NBR.astype(int)
rd.head(1)

PRODUCT_ID          NDC  CMF_PPK_NBR  CMF_PROD_NBR CMF_PACK_NBR  \
0     1368418  00067434790   1621283010       1621283          010   

                         MKTED_PROD_FORMLTN_NM  \
0  SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   

                   MKTED_PROD_FORMLTN_SHORT_NM MKTED_PROD_DETL_NM  \
0  SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG        SLOW FE OTC   

  MKTED_PROD_TYP_CD MKTED_PROD_NM      PACK_LABEL_NM    UNIFM_PROD_NM  \
0                 T       SLOW FE  SLOW FE TAB 142MG  FERROUS SULFATE   

  DOSAGE_FORM_CD           DOSAGE_FORM_NM STRNT_DESC  STRNT_VALUE_NBR  \
0           TBCR  TABLET EXTENDED RELEASE     142 MG              142   

  STRNT_UOM_CD  PACK_QTY PACK_QTY_UOM_CD  PACK_SIZE_NBR PACK_SIZE_UOM_CD  \
0           MG       1.0             BOX           90.0               EA   

   TOTAL_PACK_QTY ROUTE_ADM_CD ROUTE_ADM_NM GLBL_ROUTE_FORM_CD          UPC  \
0            90.0           OR         ORAL                BAA  30067434790   

  HCPCS_CD HCPCS_DESC  DEA_CLAS_CD                DEA_CLAS_DESC  ...  \
0        0          0          0.0  DEA Class is not applicable  ...   

   OBSOLETE_DT  ACT_IND  USC_CD                USC_DESC  USC_LVL4_CD  \
0            0        A   48111  HEMATINICS, IRON ALONE        48110   

         USC_LVL4_DESC  USC_LVL3_CD USC_LVL3_DESC USC_LVL2_CD  USC_LVL2_DESC  \
0  HEMATINICS, FERROUS        48100    HEMATINICS       48000  BLOOD FACTORS   

              GPI                                           GPI_DESC  \
0  82300010000455  Ferrous Sulfate Tab CR 142 MG (45 MG Fe Equiva...   

     GPI10_CD       GPI10_DESC GPI6_CD GPI6_DESC  GPI4_CD GPI4_DESC  GPI2_CD  \
0  8230001000  FERROUS SULFATE  823000   IRON***     8230    IRON**       82   

               GPI2_DESC  GPI_THPTC_CLAS_ID   GPI_THPTC_CLAS_DESC  \
0  HEMATOPOIETIC AGENTS*                 15  HEMATOLOGICAL AGENTS   

   GENERIC_THPTC_CLAS_ID GENERIC_THPTC_CLAS_DESC   AHFS_CD          AHFS_DESC  \
0                     59       ELECT/CALORIC/H2O  20040400  IRON PREPARATIONS   

  THPTC_CLAS_ID                      THPTC_CLAS_DESC  CMF_MKTED_PROD_NM  \
0            88  HEMATINICS & BLOOD CELL STIMULATORS            SLOW FE   

  CMF_USC_CD  
0      48111  

[1 rows x 74 columns]

Drop unnecessary columns

In [7]:
finaldf=rd.drop(['MKTED_PROD_FORMLTN_NM', 'PACK_LABEL_NM', 
                'TOTAL_PACK_QTY', 'ROUTE_ADM_CD','GLBL_ROUTE_FORM_CD','DEA_CLAS_CD','HCPCS_DESC','DEA_CLAS_DESC','OTC_IND','SPCL_PROD_IND','CMBN_DRUG_IND','ORIGINTR_PROD_IND','REPACK_IND','PATN_STATUS_IND','UNIT_TYP_ID','MULTI_SRC_CD','MULTI_SRC_NM','MKTED_PROD_FORMLTN_LNCH_DT','TPTY_RECV_ADD_DT','OBSOLETE_DT','ACT_IND','USC_LVL4_CD','USC_LVL4_DESC','USC_LVL3_CD','USC_LVL2_CD','USC_LVL2_DESC','GPI6_CD','GPI6_DESC','GPI4_CD','GPI4_DESC','GPI2_CD','GPI2_DESC','AHFS_CD', 'AHFS_DESC', 
                'CMF_MKTED_PROD_NM', 'USC_CD','USC_DESC','USC_LVL3_DESC','HCPCS_CD','ROUTE_ADM_NM','PACK_QTY_UOM_CD'], \
        axis=1)

In [8]:
finaldf.head(10)

PRODUCT_ID          NDC  CMF_PPK_NBR  CMF_PROD_NBR CMF_PACK_NBR  \
0     1368418  00067434790   1621283010       1621283          010   
1     1478043  86790001960            0             0                
2     1368417  00067434760   1621283010       1621283          010   
3     1382655  00067434772   1621283010       1621283          010   
4     1478050  86790001930   1999980001       1999980          001   
5     1368416  00067434730   1621283010       1621283          010   
6     1011934  00083012530   1621283001       1621283          001   
7     1008379  00067012575   1621283001       1621283          001   
8     1008378  00067012547   1621283001       1621283          001   
9     1011933  00083012417   1621287001       1621287          001   

                         MKTED_PROD_FORMLTN_SHORT_NM      MKTED_PROD_DETL_NM  \
0        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
1        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
2        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
3        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
4        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
5        SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG             SLOW FE OTC   
6        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG             SLOW FE OTC   
7        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG             SLOW FE OTC   
8        SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG             SLOW FE OTC   
9  SLOW FE/FOLIC ACID ORAL TABLET SR 24 HR 160-0....  SLOW FE/FOLIC ACID OTC   

  MKTED_PROD_TYP_CD       MKTED_PROD_NM               UNIFM_PROD_NM  \
0                 T             SLOW FE             FERROUS SULFATE   
1                 T             SLOW FE             FERROUS SULFATE   
2                 T             SLOW FE             FERROUS SULFATE   
3                 T             SLOW FE             FERROUS SULFATE   
4                 T             SLOW FE             FERROUS SULFATE   
5                 T             SLOW FE             FERROUS SULFATE   
6                 T             SLOW FE       FERROUS SULFATE DRIED   
7                 T             SLOW FE       FERROUS SULFATE DRIED   
8                 T             SLOW FE       FERROUS SULFATE DRIED   
9                 T  SLOW FE/FOLIC ACID  FERROUS SULFATE-FOLIC ACID   

  DOSAGE_FORM_CD           DOSAGE_FORM_NM       STRNT_DESC  STRNT_VALUE_NBR  \
0           TBCR  TABLET EXTENDED RELEASE           142 MG              142   
1           TBCR  TABLET EXTENDED RELEASE           142 MG              142   
2           TBCR  TABLET EXTENDED RELEASE           142 MG              142   
3           TBCR  TABLET EXTENDED RELEASE           142 MG              142   
4           TBCR  TABLET EXTENDED RELEASE           142 MG              142   
5           TBCR  TABLET EXTENDED RELEASE           142 MG              142   
6           TBCR  TABLET EXTENDED RELEASE           160 MG              160   
7           TBCR  TABLET EXTENDED RELEASE           160 MG              160   
8           TBCR  TABLET EXTENDED RELEASE           160 MG              160   
9           TB24          TABLET SR 24 HR  160-0.4 MG/24HR                0   

  STRNT_UOM_CD  PACK_QTY  PACK_SIZE_NBR PACK_SIZE_UOM_CD          UPC  \
0           MG       1.0           90.0               EA  30067434790   
1           MG       1.0           60.0               EA  38679001960   
2           MG       1.0           60.0               EA  30067434760   
3           MG       1.0          120.0               EA  30067434772   
4           MG       1.0           30.0               EA  38679001930   
5           MG       1.0           30.0               EA  30067434730   
6           MG       1.0          100.0               EA  30083012530   
7           MG       1.0           90.0               EA  30067012575   
8           MG       1.0           30.0   

Once the datatype has been mapped properly for better computations, check the datatypes of the affected columns again

In [9]:
print(finaldf.dtypes)

PRODUCT_ID                       int64
NDC                             object
CMF_PPK_NBR                      int32
CMF_PROD_NBR                     int32
CMF_PACK_NBR                    object
MKTED_PROD_FORMLTN_SHORT_NM     object
MKTED_PROD_DETL_NM              object
MKTED_PROD_TYP_CD               object
MKTED_PROD_NM                   object
UNIFM_PROD_NM                   object
DOSAGE_FORM_CD                  object
DOSAGE_FORM_NM                  object
STRNT_DESC                      object
STRNT_VALUE_NBR                  int32
STRNT_UOM_CD                    object
PACK_QTY                       float64
PACK_SIZE_NBR                  float64
PACK_SIZE_UOM_CD                object
UPC                             object
LBLER_NM                        object
LBLER_TYP_CD                    object
LBLER_CORP_NM                   object
GPI                             object
GPI_DESC                        object
GPI10_CD                        object
GPI10_DESC               

In [10]:
finaldf.head(15)

PRODUCT_ID          NDC  CMF_PPK_NBR  CMF_PROD_NBR CMF_PACK_NBR  \
0      1368418  00067434790   1621283010       1621283          010   
1      1478043  86790001960            0             0                
2      1368417  00067434760   1621283010       1621283          010   
3      1382655  00067434772   1621283010       1621283          010   
4      1478050  86790001930   1999980001       1999980          001   
5      1368416  00067434730   1621283010       1621283          010   
6      1011934  00083012530   1621283001       1621283          001   
7      1008379  00067012575   1621283001       1621283          001   
8      1008378  00067012547   1621283001       1621283          001   
9      1011933  00083012417   1621287001       1621287          001   
10     1377214  79854007749            0             0                
11     1414537  57896063506   1999980001       1999980          001   
12     1326789  12333024902   1282410029       1282410          029   
13     1386508  11845015265   1314437002       1314437          002   
14     1212933  11845014315            0             0                

                          MKTED_PROD_FORMLTN_SHORT_NM  \
0         SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
1         SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
2         SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
3         SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
4         SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
5         SLOW FE ORAL TABLET EXTENDED RELEASE 142 MG   
6         SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG   
7         SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG   
8         SLOW FE ORAL TABLET EXTENDED RELEASE 160 MG   
9   SLOW FE/FOLIC ACID ORAL TABLET SR 24 HR 160-0....   
10      SLOW IRON ORAL TABLET EXTENDED RELEASE 160 MG   
11  SLOW MAGNESIUM/CALCIUM ORAL TABLET ENTERICCOAT...   
12  SLOW REL IRON ORAL TABLET EXTENDED RELEASE 160 MG   
13  SLOW RELEASE IRON ORAL TABLET EXTENDED RELEASE...   
14  SLOW RELEASE IRON ORAL TABLET EXTENDED RELEASE...   

            MKTED_PROD_DETL_NM MKTED_PROD_TYP_CD           MKTED_PROD_NM  \
0                  SLOW FE OTC                 T                 SLOW FE   
1                  SLOW FE OTC                 T                 SLOW FE   
2                  SLOW FE OTC                 T                 SLOW FE   
3                  SLOW FE OTC                 T                 SLOW FE   
4                  SLOW FE OTC                 T                 SLOW FE   
5                  SLOW FE OTC                 T                 SLOW FE   
6                  SLOW FE OTC                 T                 SLOW FE   
7                  SLOW FE OTC                 T                 SLOW FE   
8                  SLOW FE OTC                 T                 SLOW FE   
9       SLOW FE/FOLIC ACID OTC                 T      SLOW FE/FOLIC ACID   
10               SLOW IRON OTC                 B               SLOW IRON   
11  SLOW MAGNESIUM/CALCIUM OTC                 B  SLOW MAGNESIUM/CALCIUM   
12           SLOW REL IRON OTC                 G           SLOW REL IRON   
13       SLOW RELEASE IRON OTC                 B       SLOW RELEASE IRON   
14       SLOW RELEASE IRON OTC                 B       SLOW RELEASE IRON   

                 UNIFM_PROD_NM DOSAGE_FORM_CD           DOSAGE_FORM_NM  \
0              FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
1              FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
2              FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
3              FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
4              FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
5              FERROUS SULFATE           TBCR  TABLET EXTENDED RELEASE   
6        FERROUS SULFATE DRIED           TBCR  TABLET EXTENDED RELEASE   
7        FERROUS SULFATE DRIED           TBCR  TABLET EXTENDED RELEASE   
8        FERROUS SULFATE DRIED           TBCR  TABLET EXTENDED RELEASE 

In [11]:
print(finaldf.shape)

(471514, 33)


Filter the NDC and GPI data alone 

In [12]:
ndc_gpi=finaldf[['NDC','GPI']]
ndc_gpi.head(1000)

NDC             GPI
0    00067434790  82300010000455
1    86790001960  82300010000455
2    00067434760  82300010000455
3    00067434772  82300010000455
4    86790001930  82300010000455
5    00067434730  82300010000455
6    00083012530  82300010100410
7    00067012575  82300010100410
8    00067012547  82300010100410
9    00083012417  82994002257520
10   79854007749  82300010100410
11   57896063506  79409902480630
12   12333024902  82300010100410
13   11845015265  82300010000460
14   11845014315  82300010100410
15   37205036865  82300010100410
16   37205003965  82300005000420
17   10939034444  82300010000455
18   40985027343  82300010100405
19   10157010261  90750030004002
20   10157102601  90750030004002
21   41388144009               0
22   10157102101  92100070000920
23   04138800182               0
24   10157103101  90059902366420
25   04138809451               0
26   10157102901  92100070000920
27   00413880018               0
28   10157010062  90750030004325
29   10157010061  90750030004325
..           ...             ...
970  70030014845  41550020100320
971  70030014844  41550020100320
972  70030014846  41550020100320
973  40986002978  41550020100120
974  00113042553  41550024100320
975  41163048066  41550024100350
976  41250042553  41550024100320
977  41280020096  43993002340303
978  94441020096  43993002340303
979  94441022307  43993002340303
980  96619026049  41550020100320
981  96619031167  41550020100320
982  96619026157  43993002277420
983  49483009024  43993002800330
984  00081007261               0
985  00081007348  43993002800330
986  00081007324  43993002800330
987  00469960033  79100010002010
988  54569206800  79100010002010
989  00074492823  79100010002010
990  00467025325  79100010002010
991  55390086210  79100010002010
992  00467025301  79100010002010
993  00009024503  79100010002010
994  00009024501  79100010002010
995  51079070245  79100010002010
996  00598112352  79100040000315
997  00598112301  79100040000315
998  00598112452  79100040000325
999  00598112401  79100040000325

[1000 rows x 2 columns]

Let us check the number of zeros in each column

In [13]:
print((ndc_gpi == 0).astype(int).sum(axis=0))

NDC        0
GPI    19581
dtype: int64


Looks like there are 19581 zero values in the entire GPI column so let us remove these rows currently and proceed with data analysis

In [14]:
print(ndc_gpi.dtypes)

NDC    object
GPI    object
dtype: object


In [15]:
ndc_gpi.head(1000)

NDC             GPI
0    00067434790  82300010000455
1    86790001960  82300010000455
2    00067434760  82300010000455
3    00067434772  82300010000455
4    86790001930  82300010000455
5    00067434730  82300010000455
6    00083012530  82300010100410
7    00067012575  82300010100410
8    00067012547  82300010100410
9    00083012417  82994002257520
10   79854007749  82300010100410
11   57896063506  79409902480630
12   12333024902  82300010100410
13   11845015265  82300010000460
14   11845014315  82300010100410
15   37205036865  82300010100410
16   37205003965  82300005000420
17   10939034444  82300010000455
18   40985027343  82300010100405
19   10157010261  90750030004002
20   10157102601  90750030004002
21   41388144009               0
22   10157102101  92100070000920
23   04138800182               0
24   10157103101  90059902366420
25   04138809451               0
26   10157102901  92100070000920
27   00413880018               0
28   10157010062  90750030004325
29   10157010061  90750030004325
..           ...             ...
970  70030014845  41550020100320
971  70030014844  41550020100320
972  70030014846  41550020100320
973  40986002978  41550020100120
974  00113042553  41550024100320
975  41163048066  41550024100350
976  41250042553  41550024100320
977  41280020096  43993002340303
978  94441020096  43993002340303
979  94441022307  43993002340303
980  96619026049  41550020100320
981  96619031167  41550020100320
982  96619026157  43993002277420
983  49483009024  43993002800330
984  00081007261               0
985  00081007348  43993002800330
986  00081007324  43993002800330
987  00469960033  79100010002010
988  54569206800  79100010002010
989  00074492823  79100010002010
990  00467025325  79100010002010
991  55390086210  79100010002010
992  00467025301  79100010002010
993  00009024503  79100010002010
994  00009024501  79100010002010
995  51079070245  79100010002010
996  00598112352  79100040000315
997  00598112301  79100040000315
998  00598112452  79100040000325
999  00598112401  79100040000325

[1000 rows x 2 columns]

In [16]:
print(ndc_gpi.dtypes)

NDC    object
GPI    object
dtype: object


In [17]:
ndc_gpi.to_csv('ndc_gpi.csv',index=False,encoding=None)

Save dataframe with only non zero rows for analysis

In [18]:
final_ndc_gpi=ndc_gpi[(ndc_gpi != 0).all(1)]

In [19]:
final_ndc_gpi.head(200)

NDC             GPI
0    00067434790  82300010000455
1    86790001960  82300010000455
2    00067434760  82300010000455
3    00067434772  82300010000455
4    86790001930  82300010000455
5    00067434730  82300010000455
6    00083012530  82300010100410
7    00067012575  82300010100410
8    00067012547  82300010100410
9    00083012417  82994002257520
10   79854007749  82300010100410
11   57896063506  79409902480630
12   12333024902  82300010100410
13   11845015265  82300010000460
14   11845014315  82300010100410
15   37205036865  82300010100410
16   37205003965  82300005000420
17   10939034444  82300010000455
18   40985027343  82300010100405
19   10157010261  90750030004002
20   10157102601  90750030004002
22   10157102101  92100070000920
24   10157103101  90059902366420
26   10157102901  92100070000920
28   10157010062  90750030004325
29   10157010061  90750030004325
30   10157100601  90750030004325
31   10157100602  90750030004325
32   12843006955  90052002104310
33   12843006555  90052002104310
..           ...             ...
179  00456141090  33200040100330
180  00456140590  33200040100320
181  54569634201  33200040100330
182  54569634200  33200040100330
183  54569611900  33200040100320
184  54868594401  33200040100320
185  54868601801  33200040100330
186  54868601800  33200040100330
188  54868594400  33200040100320
192  00313041892  43995803341220
193  05388076692  78104000000100
194  25077126050  78104000000100
195  25077012602  78104000000100
196  40986001770  77108010000330
197  11845011735  77108010000330
198  11845011739  77108010000330
199  08290207501  90800000004300
200  08290207353  97350000000000
201  08290207354  97350000000000
202  08290207355  97350000000000
203  08290207320  97350000000000
204  08290207322  97350000000000
205  08290207321  97350000000000
206  49022043346  97350000000000
207  08290207516  97375010006300
208  08290007339  97350000000000
209  08290007336  97350000000000
210  08290007337  97350000000000
211  08290007338  97350000000000
212  41353080100  97100000006300

[200 rows x 2 columns]

In [20]:
print(final_ndc_gpi.shape)

(451933, 2)


Once rows with zeros are removed , check for zeros in both the columns

In [21]:
print((final_ndc_gpi == 0).astype(int).sum(axis=0))

NDC    0
GPI    0
dtype: int64


In [22]:
final_ndc_gpi

NDC             GPI
0       00067434790  82300010000455
1       86790001960  82300010000455
2       00067434760  82300010000455
3       00067434772  82300010000455
4       86790001930  82300010000455
5       00067434730  82300010000455
6       00083012530  82300010100410
7       00067012575  82300010100410
8       00067012547  82300010100410
9       00083012417  82994002257520
10      79854007749  82300010100410
11      57896063506  79409902480630
12      12333024902  82300010100410
13      11845015265  82300010000460
14      11845014315  82300010100410
15      37205036865  82300010100410
16      37205003965  82300005000420
17      10939034444  82300010000455
18      40985027343  82300010100405
19      10157010261  90750030004002
20      10157102601  90750030004002
22      10157102101  92100070000920
24      10157103101  90059902366420
26      10157102901  92100070000920
28      10157010062  90750030004325
29      10157010061  90750030004325
30      10157100601  90750030004325
31      10157100602  90750030004325
32      12843006955  90052002104310
33      12843006555  90052002104310
...             ...             ...
471484  58016070530  66100070000105
471485  58016070528  66100070000105
471486  58016070515  66100070000105
471487  58016070514  66100070000105
471488  49452547603  66100070002900
471489  49452547601  66100070002900
471490  49452547602  66100070002900
471491  60760002060  66100070000110
471492  60760002030  66100070000110
471493  60760002020  66100070000110
471494  16590018882  66100070000110
471495  60814070001  66100070000110
471496  00832011000  66100070000110
471497  00905576251  66100070000110
471498  00905576231  66100070000110
471499  00905575231  66100070000105
471500  60713003314  66100070000110
471501  63739020303  66100070000110
471502  63739020301  66100070000110
471503  63739020302  66100070000110
471504  58016066630  66100070000110
471505  58016066628  66100070000110
471506  58016066621  66100070000110
471507  58016066615  66100070000110
471508  58016066614  66100070000110
471509  58016066610  66100070000110
471510  58016066607  66100070000110
471511  58016066697  66100070000110
471512  58016066650  66100070000110
471513  58016066720  66100070000110

[451933 rows x 2 columns]

Simple lambda mapping to get ndc and gpi together

In [23]:
final_ndc_gpi['ndc,gpi'] = list(map(lambda x, y: (x,y) , final_ndc_gpi.NDC, final_ndc_gpi.GPI))
final_ndc_gpi.head(10)

C:\Users\GiridharRavindran\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


NDC             GPI                        ndc,gpi
0  00067434790  82300010000455  (00067434790, 82300010000455)
1  86790001960  82300010000455  (86790001960, 82300010000455)
2  00067434760  82300010000455  (00067434760, 82300010000455)
3  00067434772  82300010000455  (00067434772, 82300010000455)
4  86790001930  82300010000455  (86790001930, 82300010000455)
5  00067434730  82300010000455  (00067434730, 82300010000455)
6  00083012530  82300010100410  (00083012530, 82300010100410)
7  00067012575  82300010100410  (00067012575, 82300010100410)
8  00067012547  82300010100410  (00067012547, 82300010100410)
9  00083012417  82994002257520  (00083012417, 82994002257520)

Since the data is huge, let us take a subset of the entire dataset for better analysis

In [63]:
finalndc=final_ndc_gpi['NDC'].head(200)
df_finalndc=pd.DataFrame(finalndc)
df_finalndc.columns=['NDC']
df_finalndc
#finndc=finalndc.values.tolist()

NDC
0    00067434790
1    86790001960
2    00067434760
3    00067434772
4    86790001930
5    00067434730
6    00083012530
7    00067012575
8    00067012547
9    00083012417
10   79854007749
11   57896063506
12   12333024902
13   11845015265
14   11845014315
15   37205036865
16   37205003965
17   10939034444
18   40985027343
19   10157010261
20   10157102601
22   10157102101
24   10157103101
26   10157102901
28   10157010062
29   10157010061
30   10157100601
31   10157100602
32   12843006955
33   12843006555
..           ...
179  00456141090
180  00456140590
181  54569634201
182  54569634200
183  54569611900
184  54868594401
185  54868601801
186  54868601800
188  54868594400
192  00313041892
193  05388076692
194  25077126050
195  25077012602
196  40986001770
197  11845011735
198  11845011739
199  08290207501
200  08290207353
201  08290207354
202  08290207355
203  08290207320
204  08290207322
205  08290207321
206  49022043346
207  08290207516
208  08290007339
209  08290007336
210  08290007337
211  08290007338
212  41353080100

[200 rows x 1 columns]

In [65]:
df_finalndc.to_csv('finalndc.csv',index=False,encoding=None)

Convert to List datatype

In [50]:
finalndclst=finalndc.values.tolist()
finalndclst

['00067434790',
 '86790001960',
 '00067434760',
 '00067434772',
 '86790001930',
 '00067434730',
 '00083012530',
 '00067012575',
 '00067012547',
 '00083012417',
 '79854007749',
 '57896063506',
 '12333024902',
 '11845015265',
 '11845014315',
 '37205036865',
 '37205003965',
 '10939034444',
 '40985027343',
 '10157010261',
 '10157102601',
 '10157102101',
 '10157103101',
 '10157102901',
 '10157010062',
 '10157010061',
 '10157100601',
 '10157100602',
 '12843006955',
 '12843006555',
 '12843005655',
 '12843008532',
 '10157010051',
 '10157100501',
 '10157100701',
 '00228207610',
 '00228207650',
 '00349892705',
 '00641451188',
 '00641451186',
 '00349892701',
 '00349848701',
 '55289066030',
 '55289066060',
 '55289066090',
 '58864048930',
 '00440014930',
 '00440847630',
 '00440847615',
 '00440014801',
 '00440847530',
 '00440847515',
 '00440847514',
 '00440014605',
 '00440014601',
 '41163023133',
 '41163021315',
 '41163032533',
 '41163011891',
 '41163033033',
 '41163033533',
 '41163068953',
 '411630

Keep one reference file with both NDC and GPI

In [26]:
zi_ndc_gpi= final_ndc_gpi[['NDC','GPI']].head(200)
zip_ndc_gpi=dict(list(zip(zi_ndc_gpi['NDC'], zi_ndc_gpi['GPI'])))
zip_ndc_gpi

{'00067434790': '82300010000455',
 '86790001960': '82300010000455',
 '00067434760': '82300010000455',
 '00067434772': '82300010000455',
 '86790001930': '82300010000455',
 '00067434730': '82300010000455',
 '00083012530': '82300010100410',
 '00067012575': '82300010100410',
 '00067012547': '82300010100410',
 '00083012417': '82994002257520',
 '79854007749': '82300010100410',
 '57896063506': '79409902480630',
 '12333024902': '82300010100410',
 '11845015265': '82300010000460',
 '11845014315': '82300010100410',
 '37205036865': '82300010100410',
 '37205003965': '82300005000420',
 '10939034444': '82300010000455',
 '40985027343': '82300010100405',
 '10157010261': '90750030004002',
 '10157102601': '90750030004002',
 '10157102101': '92100070000920',
 '10157103101': '90059902366420',
 '10157102901': '92100070000920',
 '10157010062': '90750030004325',
 '10157010061': '90750030004325',
 '10157100601': '90750030004325',
 '10157100602': '90750030004325',
 '12843006955': '90052002104310',
 '12843006555'

In [66]:
zi_ndc_gpi

NDC             GPI
0    00067434790  82300010000455
1    86790001960  82300010000455
2    00067434760  82300010000455
3    00067434772  82300010000455
4    86790001930  82300010000455
5    00067434730  82300010000455
6    00083012530  82300010100410
7    00067012575  82300010100410
8    00067012547  82300010100410
9    00083012417  82994002257520
10   79854007749  82300010100410
11   57896063506  79409902480630
12   12333024902  82300010100410
13   11845015265  82300010000460
14   11845014315  82300010100410
15   37205036865  82300010100410
16   37205003965  82300005000420
17   10939034444  82300010000455
18   40985027343  82300010100405
19   10157010261  90750030004002
20   10157102601  90750030004002
22   10157102101  92100070000920
24   10157103101  90059902366420
26   10157102901  92100070000920
28   10157010062  90750030004325
29   10157010061  90750030004325
30   10157100601  90750030004325
31   10157100602  90750030004325
32   12843006955  90052002104310
33   12843006555  90052002104310
..           ...             ...
179  00456141090  33200040100330
180  00456140590  33200040100320
181  54569634201  33200040100330
182  54569634200  33200040100330
183  54569611900  33200040100320
184  54868594401  33200040100320
185  54868601801  33200040100330
186  54868601800  33200040100330
188  54868594400  33200040100320
192  00313041892  43995803341220
193  05388076692  78104000000100
194  25077126050  78104000000100
195  25077012602  78104000000100
196  40986001770  77108010000330
197  11845011735  77108010000330
198  11845011739  77108010000330
199  08290207501  90800000004300
200  08290207353  97350000000000
201  08290207354  97350000000000
202  08290207355  97350000000000
203  08290207320  97350000000000
204  08290207322  97350000000000
205  08290207321  97350000000000
206  49022043346  97350000000000
207  08290207516  97375010006300
208  08290007339  97350000000000
209  08290007336  97350000000000
210  08290007337  97350000000000
211  08290007338  97350000000000
212  41353080100  97100000006300

[200 rows x 2 columns]

In [67]:
zi_ndc_gpi.to_csv('zip_ndc_gpi.csv',index=False,encoding=None)

Use list comp

In [42]:
result=[]
for k in finalndclst:
    if k in zip_ndc_gpi:
        result.append((k, zip_ndc_gpi[k]))

In [43]:
result

[('00067434790', '82300010000455'),
 ('86790001960', '82300010000455'),
 ('00067434760', '82300010000455'),
 ('00067434772', '82300010000455'),
 ('86790001930', '82300010000455'),
 ('00067434730', '82300010000455'),
 ('00083012530', '82300010100410'),
 ('00067012575', '82300010100410'),
 ('00067012547', '82300010100410'),
 ('00083012417', '82994002257520'),
 ('79854007749', '82300010100410'),
 ('57896063506', '79409902480630'),
 ('12333024902', '82300010100410'),
 ('11845015265', '82300010000460'),
 ('11845014315', '82300010100410'),
 ('37205036865', '82300010100410'),
 ('37205003965', '82300005000420'),
 ('10939034444', '82300010000455'),
 ('40985027343', '82300010100405'),
 ('10157010261', '90750030004002'),
 ('10157102601', '90750030004002'),
 ('10157102101', '92100070000920'),
 ('10157103101', '90059902366420'),
 ('10157102901', '92100070000920'),
 ('10157010062', '90750030004325'),
 ('10157010061', '90750030004325'),
 ('10157100601', '90750030004325'),
 ('10157100602', '9075003000

In [45]:
final_result=pd.DataFrame(result)

In [47]:
final_result.columns=['NDC','GPI']

In [48]:
final_result

NDC             GPI
0    00067434790  82300010000455
1    86790001960  82300010000455
2    00067434760  82300010000455
3    00067434772  82300010000455
4    86790001930  82300010000455
5    00067434730  82300010000455
6    00083012530  82300010100410
7    00067012575  82300010100410
8    00067012547  82300010100410
9    00083012417  82994002257520
10   79854007749  82300010100410
11   57896063506  79409902480630
12   12333024902  82300010100410
13   11845015265  82300010000460
14   11845014315  82300010100410
15   37205036865  82300010100410
16   37205003965  82300005000420
17   10939034444  82300010000455
18   40985027343  82300010100405
19   10157010261  90750030004002
20   10157102601  90750030004002
21   10157102101  92100070000920
22   10157103101  90059902366420
23   10157102901  92100070000920
24   10157010062  90750030004325
25   10157010061  90750030004325
26   10157100601  90750030004325
27   10157100602  90750030004325
28   12843006955  90052002104310
29   12843006555  90052002104310
..           ...             ...
170  00456141090  33200040100330
171  00456140590  33200040100320
172  54569634201  33200040100330
173  54569634200  33200040100330
174  54569611900  33200040100320
175  54868594401  33200040100320
176  54868601801  33200040100330
177  54868601800  33200040100330
178  54868594400  33200040100320
179  00313041892  43995803341220
180  05388076692  78104000000100
181  25077126050  78104000000100
182  25077012602  78104000000100
183  40986001770  77108010000330
184  11845011735  77108010000330
185  11845011739  77108010000330
186  08290207501  90800000004300
187  08290207353  97350000000000
188  08290207354  97350000000000
189  08290207355  97350000000000
190  08290207320  97350000000000
191  08290207322  97350000000000
192  08290207321  97350000000000
193  49022043346  97350000000000
194  08290207516  97375010006300
195  08290007339  97350000000000
196  08290007336  97350000000000
197  08290007337  97350000000000
198  08290007338  97350000000000
199  41353080100  97100000006300

[200 rows x 2 columns]